app/controllers/spree/admin/products_controller_decorator.rb

```diff
@@ -119,4 +131,22 @@ def strip_new_properties
       end
     end
   end
+
+  def delete_stock_params_and_set_after
+    on_demand = params[:product].delete(:on_demand)
+    on_hand = params[:product].delete(:on_hand)
+
+    yield
+
+    set_stock_levels(@product, on_hand, on_demand) if @product.errors.empty?
```

Comment:
You could use [valid?](https://apidock.com/rails/v3.2.13/ActiveResource/Validations/valid%3F) instead. It does this check for us.